diff -ruN a/arch/arm64/boot/dts/renesas/rzg2l-smarc.dtsi b/arch/arm64/boot/dts/renesas/rzg2l-smarc.dtsi
--- a/arch/arm64/boot/dts/renesas/rzg2l-smarc.dtsi	2026-07-03 04:29:34.371995142 +0000
+++ b/arch/arm64/boot/dts/renesas/rzg2l-smarc.dtsi	2026-07-03 04:31:17.777225995 +0000
@@ -140,6 +140,7 @@
 
 	port-num = <2>; /* SCIF2 */
 
+	rs485-gpio = <&pinctrl RZG2L_GPIO(48, 4) GPIO_ACTIVE_LOW>;
 	status = "okay";
 };
 
diff -ruN a/drivers/tty/serial/serial_core.c b/drivers/tty/serial/serial_core.c
--- a/drivers/tty/serial/serial_core.c	2026-01-09 05:25:58.000000000 +0000
+++ b/drivers/tty/serial/serial_core.c	2026-07-03 04:31:17.777225995 +0000
@@ -1467,6 +1467,16 @@
 	if (ret != -ENOIOCTLCMD)
 		goto out;
 
+	switch(cmd) {
+		case TIOCSSWRTS:
+			uport = state->uart_port;
+			ret = uport->ops->swrts_ctl ? uport->ops->swrts_ctl(uport, arg) : 0;
+			goto out;
+
+		default:
+			break;
+	}
+
 	mutex_lock(&port->mutex);
 	uport = uart_port_check(state);
 
@@ -2589,6 +2599,12 @@
 	.shutdown	= uart_tty_port_shutdown,
 };
 
+int is_serial_core_tty(struct tty_struct *tty)
+{
+	return ((tty->ops == &uart_ops) && tty->ops->ioctl(tty, TIOCSSWRTS, SWRTS_GET_I));
+}
+EXPORT_SYMBOL_GPL(is_serial_core_tty);
+
 /**
  *	uart_register_driver - register a driver with the uart core layer
  *	@drv: low level driver structure
diff -ruN a/drivers/tty/serial/sh-sci.c b/drivers/tty/serial/sh-sci.c
--- a/drivers/tty/serial/sh-sci.c	2026-07-03 04:29:34.375995192 +0000
+++ b/drivers/tty/serial/sh-sci.c	2026-07-03 04:31:17.777225995 +0000
@@ -49,6 +49,9 @@
 #include <linux/timer.h>
 #include <linux/tty.h>
 #include <linux/tty_flip.h>
+#include <linux/gpio.h>
+#include <linux/of_gpio.h>
+#include <linux/semaphore.h>
 
 #ifdef CONFIG_SUPERH
 #include <asm/sh_bios.h>
@@ -122,6 +125,8 @@
 	int port_type; /* PORT_VSCI or PORT_VSCIF */
 	int port_num; /* SCI or SCIF port number(physical) */
 	int tx_start; /* 0: VSCIx port is not working(TX). 1: VSCIx port is working(TX) */
+	int rs485_gpio;
+	struct semaphore rs485_sema;
 	struct vsci_device vdev;
 };
 
@@ -2258,6 +2263,10 @@
 		if(!uart_circ_empty(xmit)) {
 			vsci_transmit(sp);
 		} else {
+			if(gpio_is_valid(sp->vi->rs485_gpio)) {
+				gpio_set_value(sp->vi->rs485_gpio, 0); // RS485 INPUT mode
+				up(&sp->vi->rs485_sema);
+			}
 			sp->vi->tx_start = 0;
 		}
 
@@ -2484,6 +2493,11 @@
 
 		s->vi->tx_start = 0;
 
+		if(gpio_is_valid(s->vi->rs485_gpio)) {
+			gpio_set_value(s->vi->rs485_gpio, 0); // RS-485 INPUT mode
+			sema_init(&s->vi->rs485_sema, 0);
+		}
+
 		vsci_send_cmd(
 			&s->vi->vdev,
 			vcmd_open(s->vi->vdev.devname)
@@ -3191,6 +3205,50 @@
 	return 0;
 }
 
+static int vsci_swrts_ctl(struct uart_port *port, unsigned long arg)
+{
+	//struct circ_buf *xmit = &port->state->xmit;
+	struct sci_port *sport = to_sci_port(port);
+	int cmd, ret = 0;
+
+	/* for VSCI device only */
+	if(!sport->vi)
+		return 0;
+
+	cmd = arg & 0xff;
+
+	switch(cmd)
+	{
+		case SWRTS_GET_I: /* check rs485-sw-rts device */
+
+			return gpio_is_valid(sport->vi->rs485_gpio);
+
+		case SWRTS_SET_H:
+
+			gpio_set_value(sport->vi->rs485_gpio, 1);
+
+			break;
+
+		case SWRTS_SET_L:
+
+			/* Reset RTS GPIO to Low in SCIF TX Empty interrupt, not here */
+
+			if(down_interruptible(&sport->vi->rs485_sema)) {
+				ret = -ERESTART;
+				break;
+			}
+
+			break;
+
+		default:
+
+			dev_err(port->dev, "unsupported cmd 0x%lx for %s\n", arg, __func__);
+			return -EPERM;
+	}
+
+	return ret;
+}
+
 static const struct uart_ops sci_uart_ops = {
 	.tx_empty	= sci_tx_empty,
 	.set_mctrl	= sci_set_mctrl,
@@ -3214,6 +3272,7 @@
 	.poll_get_char	= sci_poll_get_char,
 	.poll_put_char	= sci_poll_put_char,
 #endif
+	.swrts_ctl = vsci_swrts_ctl
 };
 
 static int sci_init_clocks(struct sci_port *sci_port, struct device *dev)
@@ -3772,6 +3831,12 @@
 		if(of_property_read_u32(np, "port-num", &sp->vi->port_num)) {
 			return ERR_PTR(-EPERM);
 		}
+
+		/* check RS-485 support, optional */
+		sp->vi->rs485_gpio = of_get_named_gpio(np, "rs485-gpio", 0);
+
+		if(gpio_is_valid(sp->vi->rs485_gpio))
+			gpio_direction_output(sp->vi->rs485_gpio, 0); // output LOW(RS-485 INPUT mode) first
 	}
 
 	return p;
diff -ruN a/drivers/tty/tty_io.c b/drivers/tty/tty_io.c
--- a/drivers/tty/tty_io.c	2026-01-09 05:25:58.000000000 +0000
+++ b/drivers/tty/tty_io.c	2026-07-03 04:31:17.777225995 +0000
@@ -515,6 +515,7 @@
 static struct file *redirect;
 
 extern void tty_sysctl_init(void);
+extern int is_serial_core_tty(struct tty_struct *tty);
 
 /**
  *	tty_wakeup	-	request more data
@@ -969,9 +970,12 @@
 	struct file *file,
 	struct iov_iter *from)
 {
-	size_t count = iov_iter_count(from);
+	size_t count = iov_iter_count(from), par;
 	ssize_t ret, written = 0;
 	unsigned int chunk;
+	int serial_core_tty;
+
+	serial_core_tty = is_serial_core_tty(tty);
 
 	ret = tty_write_lock(tty, file->f_flags & O_NDELAY);
 	if (ret < 0)
@@ -1016,6 +1020,14 @@
 		tty->write_buf = buf_chunk;
 	}
 
+	if(serial_core_tty) {
+		par = count << 8;
+		if(tty->ops->ioctl(tty, TIOCSSWRTS, SWRTS_SET_H | par)) {
+			ret = -ERESTART;
+			goto out;
+		}
+	}
+
 	/* Do the write .. */
 	for (;;) {
 		size_t size = count;
@@ -1046,6 +1058,14 @@
 			break;
 		cond_resched();
 	}
+
+	if(serial_core_tty) {
+		if(tty->ops->ioctl(tty, TIOCSSWRTS, SWRTS_SET_L)) {
+			ret = -ERESTART;
+			goto out;
+		}
+	}
+
 	if (written) {
 		tty_update_time(&file_inode(file)->i_mtime);
 		ret = written;
diff -ruN a/include/linux/serial_core.h b/include/linux/serial_core.h
--- a/include/linux/serial_core.h	2026-01-09 05:25:58.000000000 +0000
+++ b/include/linux/serial_core.h	2026-07-03 04:31:17.777225995 +0000
@@ -80,6 +80,7 @@
 	void		(*poll_put_char)(struct uart_port *, unsigned char);
 	int		(*poll_get_char)(struct uart_port *);
 #endif
+	int		(*swrts_ctl)(struct uart_port *, unsigned long);
 };
 
 #define NO_POLL_CHAR		0x00ff0000
diff -ruN a/include/uapi/asm-generic/ioctls.h b/include/uapi/asm-generic/ioctls.h
--- a/include/uapi/asm-generic/ioctls.h	2026-01-09 05:25:58.000000000 +0000
+++ b/include/uapi/asm-generic/ioctls.h	2026-07-03 04:31:17.777225995 +0000
@@ -106,6 +106,18 @@
 # define FIOQSIZE	0x5460
 #endif
 
+/*
+ * Add ioctl commands for software RS-485 RTS control
+ */
+#define TIOCSSWRTS	0x5500
+enum
+{
+	SWRTS_GET_I = 0, /* Get software RS-485 RTS device support info */
+	SWRTS_SET_H, /* Set RTS GPIO to HIGH */
+	SWRTS_SET_L /* Set RTS GPIO to LOW */
+};
+
+
 /* Used for packet mode */
 #define TIOCPKT_DATA		 0
 #define TIOCPKT_FLUSHREAD	 1
