diff --git a/arch/arm64/boot/dts/renesas/r9a07g044.dtsi b/arch/arm64/boot/dts/renesas/r9a07g044.dtsi
index 6b1c77cd8261..6d68ee5531c1 100644
--- a/arch/arm64/boot/dts/renesas/r9a07g044.dtsi
+++ b/arch/arm64/boot/dts/renesas/r9a07g044.dtsi
@@ -497,6 +497,44 @@ sci1: serial@1004d400 {
 			status = "disabled";
 		};
 
+		mhu: mhu@10400000 {
+			compatible = "renesas,mhu-r9a07g044";
+				reg = <0 0x10400000 0 0x1000>, /* MHU reg */
+				      <0 0x00020000 0 0x4000>; /* Shared memory of Linux side */
+			interrupts = <GIC_SPI 72 IRQ_TYPE_LEVEL_HIGH>,/* CH4 MSG, M33->CORE0 */
+				     <GIC_SPI 75 IRQ_TYPE_LEVEL_HIGH>,/* CH1 RSP, M33->CORE0 */
+				     <GIC_SPI 73 IRQ_TYPE_LEVEL_HIGH>,/* CH5 MSG, M33->CORE1 */
+				     <GIC_SPI 77 IRQ_TYPE_LEVEL_HIGH>;/* CH3 RSP, M33->CORE1 */
+			interrupt-names = "msg4-core0", "rsp1-core0",
+					"msg5-core1", "rsp3-core1";
+
+			/* RX INTR CH, TX INTR CH, CMD CH, Bit16 = 1 means RSP CH */
+			port-0 = <0x0004 0x10001 0x0001>;
+			port-1 = <0x0005 0x10003 0x0003>;
+
+			shm-rtos-base = <0x20020000>; /* Shared memory base of RTOS side */
+
+			status = "disabled";
+		};
+
+		vscig0: serial@0000 {
+			compatible = "renesas,vscig-r9a07g044";
+
+			status = "disabled";
+		};
+
+		vscif1: serial@0001 {
+			compatible = "renesas,vscif-r9a07g044";
+
+			status = "disabled";
+		};
+
+		vscif2: serial@0002 {
+			compatible = "renesas,vscif-r9a07g044";
+
+			status = "disabled";
+		};
+
 		canfd: can@10050000 {
 			compatible = "renesas,r9a07g044-canfd", "renesas,rzg2l-canfd";
 			reg = <0 0x10050000 0 0x8000>;
diff --git a/arch/arm64/boot/dts/renesas/rzg2l-smarc-pinfunction.dtsi b/arch/arm64/boot/dts/renesas/rzg2l-smarc-pinfunction.dtsi
index e9f244c33d55..3299d67ce2a7 100644
--- a/arch/arm64/boot/dts/renesas/rzg2l-smarc-pinfunction.dtsi
+++ b/arch/arm64/boot/dts/renesas/rzg2l-smarc-pinfunction.dtsi
@@ -85,6 +85,21 @@ scif2_pins: scif2 {
 			 <RZG2L_PORT_PINMUX(48, 4, 1)>; /* RTS# */
 	};
 
+	vscif1_pins: vscif1 {
+		pinmux = <RZG2L_PORT_PINMUX(40, 0, 1)>, /* TxD */
+			<RZG2L_PORT_PINMUX(40, 1, 1)>; /* RxD */
+	};
+
+	vscif2_pins: vscif2 {
+		pinmux = <RZG2L_PORT_PINMUX(48, 0, 1)>, /* TxD */
+			<RZG2L_PORT_PINMUX(48, 1, 1)>; /* RxD */
+	};
+
+	vscig0_pins: vscig0 {
+		pinmux = <RZG2L_PORT_PINMUX(40, 0, 5)>, /* RxD */
+			<RZG2L_PORT_PINMUX(40, 1, 5)>; /* TxD */
+	};
+
 	sd1-pwr-en-hog {
 		gpio-hog;
 		gpios = <RZG2L_GPIO(39, 2) GPIO_ACTIVE_HIGH>;
diff --git a/arch/arm64/boot/dts/renesas/rzg2l-smarc-som.dtsi b/arch/arm64/boot/dts/renesas/rzg2l-smarc-som.dtsi
index 21cf198b3c17..e37b24fdc203 100644
--- a/arch/arm64/boot/dts/renesas/rzg2l-smarc-som.dtsi
+++ b/arch/arm64/boot/dts/renesas/rzg2l-smarc-som.dtsi
@@ -165,11 +165,11 @@ raa215300: pmic@12 {
 };
 
 &ostm1 {
-	status = "okay";
+	status = "disabled";
 };
 
 &ostm2 {
-	status = "okay";
+	status = "disabled";
 };
 
 &pinctrl {
diff --git a/arch/arm64/boot/dts/renesas/rzg2l-smarc.dtsi b/arch/arm64/boot/dts/renesas/rzg2l-smarc.dtsi
index ee3d96fdb616..c1f8484de32a 100644
--- a/arch/arm64/boot/dts/renesas/rzg2l-smarc.dtsi
+++ b/arch/arm64/boot/dts/renesas/rzg2l-smarc.dtsi
@@ -10,7 +10,8 @@
 
 / {
 	aliases {
-		serial1 = &scif2;
+		serial1 = &vscig0;
+		serial2 = &vscif2;
 		i2c3 = &i2c3;
 	};
 
@@ -172,10 +173,41 @@ &scif2 {
 	pinctrl-names = "default";
 
 	uart-has-rtscts;
-	status = "okay";
+	status = "disabled";
 };
 #endif
 
+&mhu {
+	status = "okay";
+};
+
+&vscig0 {
+	pinctrl-0 = <&vscig0_pins>;
+	pinctrl-names = "default";
+
+	port-num = <0>; /* SCIg0 */
+
+	status = "okay";
+};
+
+&vscif1 {
+	pinctrl-0 = <&vscif1_pins>;
+	pinctrl-names = "default";
+
+	port-num = <1>; /* SCIF1 */
+
+	status = "disabled";
+};
+
+&vscif2 {
+	pinctrl-0 = <&vscif2_pins>;
+	pinctrl-names = "default";
+
+	port-num = <2>; /* SCIF2 */
+
+	status = "okay";
+};
+
 &ssi0 {
 	pinctrl-0 = <&ssi0_pins>;
 	pinctrl-names = "default";
diff --git a/drivers/clk/renesas/r9a07g044-cpg.c b/drivers/clk/renesas/r9a07g044-cpg.c
index 77ca3a789568..cea59792e667 100644
--- a/drivers/clk/renesas/r9a07g044-cpg.c
+++ b/drivers/clk/renesas/r9a07g044-cpg.c
@@ -487,6 +487,11 @@ static const unsigned int r9a07g044_crit_mod_clks[] __initconst = {
 	MOD_CLK_BASE + R9A07G044_GIC600_GICCLK,
 	MOD_CLK_BASE + R9A07G044_IA55_CLK,
 	MOD_CLK_BASE + R9A07G044_DMAC_ACLK,
+	MOD_CLK_BASE + R9A07G044_SCI0_CLKP,/* Optional, VSCIg device */
+	MOD_CLK_BASE + R9A07G044_SCIF2_CLK_PCK, /* Optional, VSCIF device */
+	MOD_CLK_BASE + R9A07G044_OSTM1_PCLK, /* Must. Don't change. Timer for CM33 */
+	MOD_CLK_BASE + R9A07G044_OSTM2_PCLK, /* Must. Don't change, Timer for CM33 */
+	MOD_CLK_BASE + R9A07G044_SCIF1_CLK_PCK, /* Optional, VSCIF device */
 };
 
 static const unsigned int r9a07g044_no_pm_mod_clks[] = {
diff --git a/drivers/tty/serial/Kconfig b/drivers/tty/serial/Kconfig
index 28e4beeabf8f..0c7d947c4377 100644
--- a/drivers/tty/serial/Kconfig
+++ b/drivers/tty/serial/Kconfig
@@ -638,9 +638,13 @@ config SERIAL_IP22_ZILOG_CONSOLE
 	depends on SERIAL_IP22_ZILOG=y
 	select SERIAL_CORE_CONSOLE
 
+config MHU
+	bool "Message Handling Unit support"
+	default y
+
 config SERIAL_SH_SCI
 	tristate "SuperH SCI(F) serial port support"
-	depends on SUPERH || ARCH_RENESAS || COMPILE_TEST
+	depends on MHU && (SUPERH || ARCH_RENESAS || H8300 || COMPILE_TEST)
 	select SERIAL_CORE
 	select SERIAL_MCTRL_GPIO if GPIOLIB
 
diff --git a/drivers/tty/serial/Makefile b/drivers/tty/serial/Makefile
index 6ff74f0a9530..63c193ef0524 100644
--- a/drivers/tty/serial/Makefile
+++ b/drivers/tty/serial/Makefile
@@ -78,7 +78,8 @@ obj-$(CONFIG_SERIAL_SCCNXP)		+= sccnxp.o
 obj-$(CONFIG_SERIAL_SC16IS7XX)		+= sc16is7xx.o
 obj-$(CONFIG_SERIAL_SC16IS7XX_SPI)	+= sc16is7xx_spi.o
 obj-$(CONFIG_SERIAL_SC16IS7XX_I2C)	+= sc16is7xx_i2c.o
-obj-$(CONFIG_SERIAL_SH_SCI)		+= sh-sci.o
+obj-$(CONFIG_SERIAL_SH_SCI)		+= sh-sci.o sh-vsci.o
+obj-$(CONFIG_MHU) += mhu.o
 obj-$(CONFIG_SERIAL_SIFIVE)		+= sifive.o
 obj-$(CONFIG_SERIAL_SPRD)		+= sprd_serial.o
 obj-$(CONFIG_SERIAL_ST_ASC)		+= st-asc.o
diff --git a/drivers/tty/serial/sh-sci.c b/drivers/tty/serial/sh-sci.c
index d4e7eb32abab..644d9bbfde4f 100644
--- a/drivers/tty/serial/sh-sci.c
+++ b/drivers/tty/serial/sh-sci.c
@@ -57,6 +57,8 @@
 #include "serial_mctrl_gpio.h"
 #include "sh-sci.h"
 
+#include "sh-vsci.h"
+
 /* Offsets into the sci_port->irqs array */
 enum {
 	SCIx_ERI_IRQ,
@@ -128,6 +130,13 @@ struct sci_port_params {
 	unsigned int error_clear;
 };
 
+struct vsci_info {
+	int port_type; /* PORT_VSCI or PORT_VSCIF */
+	int port_num; /* SCI or SCIF port number(physical) */
+	int tx_start; /* 0: VSCIx port is not working(TX). 1: VSCIx port is working(TX) */
+	struct vsci_device vdev;
+};
+
 struct sci_port {
 	struct uart_port	port;
 
@@ -175,6 +184,8 @@ struct sci_port {
 	bool has_rtscts;
 	bool autorts;
 	bool tx_occurred;
+
+	struct vsci_info *vi;
 };
 
 #define SCI_NPORTS CONFIG_SERIAL_SH_SCI_NR_UARTS
@@ -588,6 +599,9 @@ static void sci_port_enable(struct sci_port *sci_port)
 
 	pm_runtime_get_sync(sci_port->port.dev);
 
+	if(sci_port->vi)
+		return;
+
 	for (i = 0; i < SCI_NUM_CLKS; i++) {
 		clk_prepare_enable(sci_port->clks[i]);
 		sci_port->clk_rates[i] = clk_get_rate(sci_port->clks[i]);
@@ -602,8 +616,10 @@ static void sci_port_disable(struct sci_port *sci_port)
 	if (!sci_port->port.dev)
 		return;
 
-	for (i = SCI_NUM_CLKS; i-- > 0; )
-		clk_disable_unprepare(sci_port->clks[i]);
+	if(!sci_port->vi) {
+		for (i = SCI_NUM_CLKS; i-- > 0; )
+			clk_disable_unprepare(sci_port->clks[i]);
+	}
 
 	pm_runtime_put_sync(sci_port->port.dev);
 }
@@ -620,11 +636,109 @@ static inline unsigned long port_rx_irq_mask(struct uart_port *port)
 	return SCSCR_RIE | (to_sci_port(port)->cfg->scscr & SCSCR_REIE);
 }
 
+static void vsci_receive(struct sci_port *sp, int bytes, int push)
+{
+	struct uart_port *port = &sp->port;
+	struct tty_port *tport = &port->state->port;
+	struct vsci_device *vd = &sp->vi->vdev;
+	unsigned char *pb;
+	uint64_t *buf, val;
+	int count, c;
+
+	pb = (unsigned char *)vd->rbuf;
+
+	if(bytes > (VSCI_BUF_SIZE / 2)) {
+		pb += VSCI_BUF_SIZE / 2;
+		bytes -= VSCI_BUF_SIZE / 2;
+	}
+
+	buf = (uint64_t *)pb;
+	count = bytes;
+
+	while(bytes > 0) {
+		val = *buf++;
+		pb = (unsigned char *)&val;
+		c = (bytes >= 8) ? 8 : bytes;
+		bytes -= c;
+
+		while(c) {
+			if(uart_handle_sysrq_char(port, *pb)) {
+				c--;
+				pb++;
+				count--;
+				continue;
+			}
+
+			tty_insert_flip_char(tport, *pb, TTY_NORMAL);
+			c--;
+			pb++;
+		}
+	}
+
+	port->icount.rx += count;
+
+	if(push)
+		tty_flip_buffer_push(tport);
+}
+
+static void vsci_transmit(struct sci_port *sp)
+{
+	struct uart_port *port = &sp->port;
+	struct tty_port *tport = &port->state->port;
+	struct vsci_device *vd = &sp->vi->vdev;
+	unsigned char *pb;
+	int bytes, count, actual_count = 0;
+	uint64_t *buf, val;
+	int i;
+
+	buf = (uint64_t *)vd->tbuf;
+	count = kfifo_len(&tport->xmit_fifo);
+	bytes = (count >= VSCI_BUF_SIZE) ? VSCI_BUF_SIZE : count;
+
+	/* Read data from kfifo in 8-byte chunks */
+	while(bytes > 0) {
+		val = 0;
+		pb = (unsigned char *)&val;
+		i = (bytes >= 8) ? 8 : bytes;
+		bytes -= i;
+
+		/* Get each byte from kfifo */
+		while(i > 0) {
+			if (kfifo_get(&tport->xmit_fifo, pb)) {
+				pb++;
+				actual_count++;
+			} else {
+				break;
+			}
+			i--;
+		}
+
+		*buf++ = val;
+	}
+
+	port->icount.tx += actual_count;
+
+	if (actual_count > 0)
+		vsci_send_cmd(vd, vcmd_txd_rdy(actual_count));
+}
+
 static void sci_start_tx(struct uart_port *port)
 {
 	struct sci_port *s = to_sci_port(port);
 	unsigned short ctrl;
 
+	if(s->vi) {
+		if(s->vi->tx_start)
+			return;
+
+		if(!kfifo_is_empty(&s->port.state->port.xmit_fifo)) {  /* Changed */
+			s->vi->tx_start = 1;
+			vsci_transmit(s);
+		}
+
+		return;
+	}
+
 #ifdef CONFIG_SERIAL_SH_SCI_DMA
 	if (port->type == PORT_SCIFA || port->type == PORT_SCIFB) {
 		u16 new, scr = sci_serial_in(port, SCSCR);
@@ -667,6 +781,12 @@ static void sci_start_tx(struct uart_port *port)
 static void sci_stop_tx(struct uart_port *port)
 {
 	unsigned short ctrl;
+	struct sci_port *s = to_sci_port(port);
+
+	if(s->vi) {
+		vsci_send_cmd(&s->vi->vdev, vcmd_stop(0, 1));
+		return;
+	}
 
 	/* Clear TIE (Transmit Interrupt Enable) bit in SCSCR */
 	ctrl = sci_serial_in(port, SCSCR);
@@ -690,6 +810,12 @@ static void sci_stop_tx(struct uart_port *port)
 static void sci_start_rx(struct uart_port *port)
 {
 	unsigned short ctrl;
+	struct sci_port *s = to_sci_port(port);
+
+	if(s->vi) {
+		vsci_send_cmd(&s->vi->vdev, vcmd_start(1, 0));
+		return;
+	}
 
 	ctrl = sci_serial_in(port, SCSCR) | port_rx_irq_mask(port);
 
@@ -702,6 +828,12 @@ static void sci_start_rx(struct uart_port *port)
 static void sci_stop_rx(struct uart_port *port)
 {
 	unsigned short ctrl;
+	struct sci_port *s = to_sci_port(port);
+
+	if(s->vi) {
+		vsci_send_cmd(&s->vi->vdev, vcmd_stop(1, 0));
+		return;
+	}
 
 	ctrl = sci_serial_in(port, SCSCR);
 
@@ -1174,6 +1306,9 @@ static ssize_t rx_fifo_trigger_show(struct device *dev,
 	struct uart_port *port = dev_get_drvdata(dev);
 	struct sci_port *sci = to_sci_port(port);
 
+	if(sci->vi)
+		return -EPERM;
+
 	return sprintf(buf, "%d\n", sci->rx_trigger);
 }
 
@@ -1186,6 +1321,9 @@ static ssize_t rx_fifo_trigger_store(struct device *dev,
 	int ret;
 	long r;
 
+	if(sci->vi)
+		return -EPERM;
+
 	ret = kstrtol(buf, 0, &r);
 	if (ret)
 		return ret;
@@ -1207,6 +1345,9 @@ static ssize_t rx_fifo_timeout_show(struct device *dev,
 	struct sci_port *sci = to_sci_port(port);
 	int v;
 
+	if(sci->vi)
+		return -EPERM;
+
 	if (port->type == PORT_HSCIF)
 		v = sci->hscif_tot >> HSSCR_TOT_SHIFT;
 	else
@@ -1225,6 +1366,9 @@ static ssize_t rx_fifo_timeout_store(struct device *dev,
 	int ret;
 	long r;
 
+	if(sci->vi)
+		return -EPERM;
+
 	ret = kstrtol(buf, 0, &r);
 	if (ret)
 		return ret;
@@ -2113,12 +2257,67 @@ static void sci_free_irq(struct sci_port *port)
 	}
 }
 
+static int vsci_intr_cb_rx(uint32_t msg, void *arg)
+{
+	struct vsci_device *vd = container_of(arg, struct vsci_device, mp);
+	struct sci_port *sp = (struct sci_port *)vd->sciport;
+	uint32_t opcode = VSCI_GET_OPCODE(msg);
+	union vscir_rx_rdy r;
+	union vscir_rxd_rdy rd;
+	int bytes;
+
+	if(VSCIR_RX_RDY == opcode) {
+		r.d = msg;
+		vsci_receive(sp, r.r.bytes, 1);
+	} else if(VSCIR_RXD_RDY == opcode) {
+		rd.d = msg;
+		bytes = (rd.r.bytes & ((VSCI_BUF_SIZE / 2) - 1)) ? rd.r.bytes : 0;
+		if(bytes)
+			vsci_receive(sp, bytes, 1);
+	} else {
+		dev_err(vd->platdev, "unsupported VSCI REQ 0x%08x was found in %s\n", msg, __func__);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int vsci_intr_cb_tx(uint32_t msg, void *arg)
+{
+	struct vsci_device *vd = container_of(arg, struct vsci_device, mp);
+	struct sci_port *sp = (struct sci_port *)vd->sciport;
+	struct tty_port *tport = &sp->port.state->port;  /* Changed */
+	uint32_t opcode = VSCI_GET_OPCODE(msg);
+
+	if(VSCIR_TX_END == opcode) {
+		if(!kfifo_is_empty(&tport->xmit_fifo)) {  /* Changed */
+			vsci_transmit(sp);
+		} else {
+			sp->vi->tx_start = 0;
+		}
+	} else {
+		dev_err(vd->platdev, "unsupported VSCI REQ 0x%08x was found in %s\n", msg, __func__);
+		return -EIO;
+	}
+
+	if(kfifo_len(&tport->xmit_fifo) < WAKEUP_CHARS)  /* Changed */
+		uart_write_wakeup(&sp->port);
+
+	return 0;
+}
+
 static unsigned int sci_tx_empty(struct uart_port *port)
 {
-	unsigned short status = sci_serial_in(port, SCxSR);
-	unsigned short in_tx_fifo = sci_txfill(port);
+	unsigned short status;
+	unsigned short in_tx_fifo;
 	struct sci_port *s = to_sci_port(port);
 
+	if(s->vi)
+		return !s->vi->tx_start ? TIOCSER_TEMT : 0;
+
+	status = sci_serial_in(port, SCxSR);
+	in_tx_fifo = sci_txfill(port);
+
 	sci_dma_check_tx_occurred(s);
 
 	if (!s->tx_occurred)
@@ -2183,6 +2382,9 @@ static void sci_set_mctrl(struct uart_port *port, unsigned int mctrl)
 {
 	struct sci_port *s = to_sci_port(port);
 
+	if(s->vi)
+		return;
+
 	if (mctrl & TIOCM_LOOP) {
 		const struct plat_sci_reg *reg;
 
@@ -2231,6 +2433,9 @@ static unsigned int sci_get_mctrl(struct uart_port *port)
 	struct mctrl_gpios *gpios = s->gpios;
 	unsigned int mctrl = 0;
 
+	if(s->vi)
+		return 0;
+
 	mctrl_gpio_get(gpios, &mctrl);
 
 	/*
@@ -2253,6 +2458,11 @@ static unsigned int sci_get_mctrl(struct uart_port *port)
 
 static void sci_enable_ms(struct uart_port *port)
 {
+	struct sci_port *s = to_sci_port(port);  /* Add this line */
+
+	if(s->vi)
+		return;
+
 	mctrl_gpio_enable_ms(to_sci_port(port)->gpios);
 }
 
@@ -2260,6 +2470,10 @@ static void sci_break_ctl(struct uart_port *port, int break_state)
 {
 	unsigned short scscr, scsptr;
 	unsigned long flags;
+	struct sci_port *s = to_sci_port(port);
+
+	if(s->vi)
+		return;
 
 	/* check whether the port has SCSPTR */
 	if (!sci_getreg(port, SCSPTR)->size) {
@@ -2292,6 +2506,27 @@ static int sci_startup(struct uart_port *port)
 	struct sci_port *s = to_sci_port(port);
 	int ret;
 
+	if(s->vi) {
+		ret = vsci_alloc_device(
+			port->dev,
+			&s->vi->vdev,
+			(void *)s,
+			s->vi->port_type,
+			s->vi->port_num,
+			vsci_intr_cb_rx,
+			vsci_intr_cb_tx
+		);
+
+		if(-1 == ret)
+			return -ENODEV;
+
+		s->vi->tx_start = 0;
+
+		vsci_send_cmd(&s->vi->vdev, vcmd_open(s->vi->vdev.devname));
+
+		return 0;
+	}
+
 	dev_dbg(port->dev, "%s(%d)\n", __func__, port->line);
 
 	s->tx_occurred = false;
@@ -2312,6 +2547,18 @@ static void sci_shutdown(struct uart_port *port)
 	unsigned long flags;
 	u16 scr;
 
+	if(s->vi) {
+		sci_stop_rx(port);
+		sci_stop_tx(port);
+
+		s->vi->tx_start = 0;
+		vsci_send_cmd(&s->vi->vdev, vcmd_close());
+
+		vsci_free_device(&s->vi->vdev);
+
+		return;
+	}
+
 	dev_dbg(port->dev, "%s(%d)\n", __func__, port->line);
 
 	s->autorts = false;
@@ -2475,6 +2722,9 @@ static void sci_reset(struct uart_port *port)
 	unsigned int status;
 	struct sci_port *s = to_sci_port(port);
 
+	if(s->vi)
+		return;
+
 	sci_serial_out(port, SCSCR, s->hscif_tot);	/* TE=0, RE=0, CKE1=0 */
 
 	reg = sci_getreg(port, SCFCR);
@@ -2516,6 +2766,30 @@ static void sci_set_termios(struct uart_port *port, struct ktermios *termios,
 	unsigned long max_freq = 0;
 	int best_clk = -1;
 	unsigned long flags;
+	struct vsci_device *vd;  /* Add this line */
+
+	if(s->vi) {
+		vd = &s->vi->vdev;
+
+		baud = uart_get_baud_rate(port, termios, NULL, 0, 10000000);
+
+		uart_update_timeout(port, termios->c_cflag, baud);
+
+		vsci_baud_adjust(vd, baud);
+		
+		vsci_send_cmd(vd, vcmd_conf(
+			vsci_baud_enc(baud),
+			(CS8 == (termios->c_cflag & CSIZE)) ? 8 : 9,
+			(termios->c_cflag & PARENB) ? ((PARODD & termios->c_cflag) ? VSCI_PARITY_ODD : VSCI_PARITY_EVEN) : VSCI_PARITY_OFF,
+			(termios->c_cflag & CSTOPB) ? 2 : 1,
+			s->has_rtscts ? VSCI_HW_FLOWCTRL_ON : VSCI_HW_FLOWCTRL_OFF
+		));
+
+		if ((termios->c_cflag & CREAD) != 0)
+			sci_start_rx(port);
+
+		return;
+	}
 
 	if ((termios->c_cflag & CSIZE) == CS7) {
 		smr_val |= SCSMR_CHR;
@@ -2789,6 +3063,10 @@ static const char *sci_type(struct uart_port *port)
 		return "scifb";
 	case PORT_HSCIF:
 		return "hscif";
+	case PORT_VSCIG:
+		return "vscig";
+	case PORT_VSCIF:
+		return "vscif";
 	}
 
 	return NULL;
@@ -2826,6 +3104,9 @@ static void sci_release_port(struct uart_port *port)
 {
 	struct sci_port *sport = to_sci_port(port);
 
+	if(sport->vi)
+		return;  /* Remove the '0' */
+
 	if (port->dev->of_node || (port->flags & UPF_IOREMAP)) {
 		iounmap(port->membase);
 		port->membase = NULL;
@@ -2840,6 +3121,9 @@ static int sci_request_port(struct uart_port *port)
 	struct sci_port *sport = to_sci_port(port);
 	int ret;
 
+	if(sport->vi)
+		return 0;
+
 	res = request_mem_region(port->mapbase, sport->reg_size,
 				 dev_name(port->dev));
 	if (unlikely(res == NULL)) {
@@ -2999,6 +3283,14 @@ static int sci_init_single(struct platform_device *dev,
 	port->line	= index;
 	port->has_sysrq = IS_ENABLED(CONFIG_SERIAL_SH_SCI_CONSOLE);
 
+	if(sci_port->vi) {
+		port->dev = &dev->dev;
+		port->flags = UPF_FIXED_PORT | UPF_BOOT_AUTOCONF | p->flags;
+		port->fifosize = VSCI_BUF_SIZE;
+		port->mapbase = vsci_get_mapbase(sci_port->vi->port_type, sci_port->vi->port_num);
+		return 0;
+	}
+
 	res = platform_get_resource(dev, IORESOURCE_MEM, 0);
 	if (res == NULL)
 		return -ENOMEM;
@@ -3267,6 +3559,10 @@ static void sci_remove(struct platform_device *dev)
 	struct sci_port *port = platform_get_drvdata(dev);
 	unsigned int type = port->port.type;	/* uart_remove_... clears it */
 
+	/* Add VSCI cleanup here */
+	if(port->vi)
+		kfree(port->vi);
+
 	sci_ports_in_use &= ~BIT(port->port.line);
 	uart_remove_one_port(&sci_uart_driver, &port->port);
 
@@ -3329,6 +3625,12 @@ static const struct of_device_id of_sci_match[] __maybe_unused = {
 	}, {
 		.compatible = "renesas,sci",
 		.data = SCI_OF_DATA(PORT_SCI, SCIx_SCI_REGTYPE),
+	}, {
+		.compatible = "renesas,vscig-r9a07g044",
+		.data = SCI_OF_DATA(PORT_VSCIG, 0),
+	}, {
+		.compatible = "renesas,vscif-r9a07g044",
+		.data = SCI_OF_DATA(PORT_VSCIF, 0),
 	}, {
 		/* Terminator */
 	},
@@ -3399,6 +3701,26 @@ static struct plat_sci_port *sci_parse_dt(struct platform_device *pdev,
 
 	sp->has_rtscts = of_property_read_bool(np, "uart-has-rtscts");
 
+	sp->vi = NULL;
+
+	if((PORT_VSCIG == p->type) || (PORT_VSCIF == p->type)) {
+		sp->vi = (struct vsci_info *)kzalloc(sizeof(struct vsci_info), GFP_KERNEL);
+
+		if(NULL == sp->vi) {
+			dev_err(&pdev->dev, "no memory for vsci device info\n");
+			return ERR_PTR(-ENOMEM);
+		}
+
+		sp->vi->port_type = p->type;
+
+		if(of_property_read_u32(np, "port-num", &sp->vi->port_num)) {
+			dev_err(&pdev->dev, "missing port-num property for VSCI device\n");
+			kfree(sp->vi);
+			return ERR_PTR(-EPERM);
+		}
+
+	}
+
 	return p;
 }
 
@@ -3417,6 +3739,7 @@ static int sci_probe_single(struct platform_device *dev,
 		dev_notice(&dev->dev, "Consider bumping CONFIG_SERIAL_SH_SCI_NR_UARTS!\n");
 		return -EINVAL;
 	}
+
 	BUILD_BUG_ON(SCI_NPORTS > sizeof(sci_ports_in_use) * 8);
 	if (sci_ports_in_use & BIT(index))
 		return -EBUSY;
@@ -3432,28 +3755,32 @@ static int sci_probe_single(struct platform_device *dev,
 	mutex_unlock(&sci_uart_registration_lock);
 
 	ret = sci_init_single(dev, sciport, index, p, false);
-	if (ret)
+	if (ret) {
+		dev_err(&dev->dev, "sci_init_single failed: %d\n", ret);
 		return ret;
+	}
 
 	sciport->port.dev = &dev->dev;
 	ret = devm_pm_runtime_enable(&dev->dev);
 	if (ret)
 		return ret;
 
-	sciport->gpios = mctrl_gpio_init(&sciport->port, 0);
-	if (IS_ERR(sciport->gpios))
-		return PTR_ERR(sciport->gpios);
+	if (!sciport->vi) {
+		sciport->gpios = mctrl_gpio_init(&sciport->port, 0);
+		if (IS_ERR(sciport->gpios))
+			return PTR_ERR(sciport->gpios);
 
-	if (sciport->has_rtscts) {
-		if (mctrl_gpio_to_gpiod(sciport->gpios, UART_GPIO_CTS) ||
-		    mctrl_gpio_to_gpiod(sciport->gpios, UART_GPIO_RTS)) {
-			dev_err(&dev->dev, "Conflicting RTS/CTS config\n");
-			return -EINVAL;
+		if (sciport->has_rtscts) {
+			if (mctrl_gpio_to_gpiod(sciport->gpios, UART_GPIO_CTS) ||
+				mctrl_gpio_to_gpiod(sciport->gpios, UART_GPIO_RTS)) {
+				dev_err(&dev->dev, "Conflicting RTS/CTS config\n");
+				return -EINVAL;
+			}
+			sciport->port.flags |= UPF_HARD_FLOW;
 		}
-		sciport->port.flags |= UPF_HARD_FLOW;
 	}
 
-	if (sci_uart_earlycon && sci_ports[0].port.mapbase == sci_res->start) {
+	if (sci_res && sci_uart_earlycon && sci_ports[0].port.mapbase == sci_res->start) {
 		/*
 		 * In case:
 		 * - this is the earlycon port (mapped on index 0 in sci_ports[]) and
@@ -3526,9 +3853,12 @@ static int sci_probe(struct platform_device *dev)
 	 * Except when the probed port is the same as the earlycon port.
 	 */
 
-	res = platform_get_resource(dev, IORESOURCE_MEM, 0);
-	if (!res)
-		return -ENODEV;
+	/* VSCI devices are virtual - they don't have memory resources */
+	if (!sp->vi) {
+		res = platform_get_resource(dev, IORESOURCE_MEM, 0);
+		if (!res)
+			return -ENODEV;
+	}
 
 	if (sci_uart_earlycon && sp == &sci_ports[0] && sp->port.mapbase != res->start)
 		return dev_err_probe(&dev->dev, -EBUSY, "sci_port[0] is used by earlycon!\n");
diff --git a/include/uapi/linux/serial_core.h b/include/uapi/linux/serial_core.h
index 9c007a106330..e1b7feafb78e 100644
--- a/include/uapi/linux/serial_core.h
+++ b/include/uapi/linux/serial_core.h
@@ -234,4 +234,8 @@
 /* Generic type identifier for ports which type is not important to userspace. */
 #define PORT_GENERIC	(-1)
 
+/* Renesas SH-VSCIx */
+#define PORT_VSCIG	124
+#define PORT_VSCIF	125
+
 #endif /* _UAPILINUX_SERIAL_CORE_H */
